lib/Extension/LanguageServerPhpCsFixer/Provider/PhpCsFixerDiagnosticsProvider.php

```diff
@@ -0,0 +1,274 @@
+                    $replacedLines = null;
+                    /** @var int|null */
+                    $startLineNo = null;
+
+                    foreach ($chunk->getLines() as $index => $line) {
+                        // increment orig file line number (added lines are not part of orig file)
+                        if (in_array($line->getType(), [Line::UNCHANGED, Line::REMOVED])) {
+                            $lineNo++;
+                        }
+
+                        $lastChangedLine = end($changedLines);
+
+                        // consume same as previous line
+                        if ($lastChangedLine && $line->getType() === $lastChangedLine->getType()) {
+                            $changedLines[] = $line;
+                            continue;
+                        }
+
+                        // consume lines if previous were removed and now we getting a replacement ones
+                        if ($lastChangedLine && $lastChangedLine->getType() === Line::REMOVED && $line->getType() === Line::ADDED) {
+                            $replacedLines = $changedLines;
+                            $changedLines = [$line];
+
+                            continue;
+                        }
+
+                        if ($lastChangedLine) {
+                            if (empty($changedLines) || !$startLineNo) {
+                                throw new LogicException("Missing data that's expected to be set");
+                            }
+
+                            $startPos = new Position($startLineNo, 0);
+                            $lineLength = strlen($lastChangedLine->getContent());
+                            $endPos = $lineLength
+                                ? new Position($lineNo - 1, $lineLength)
+                                : new Position($lineNo, 0);
+
+                            if ($replacedLines) {
+                                $firstLineA = $replacedLines[0]->getContent();
+                                $firstLineB = $changedLines[0]->getContent();
+                                $lastLineA = end($replacedLines)->getContent();
+                                $lastLineB = end($changedLines)->getContent();
+
+                                $startChars = StringSharedChars::startLength($firstLineA, $firstLineB);
+                                $endChars = StringSharedChars::endPos($lastLineA, $lastLineB);
+
+                                $startPos = new Position($startLineNo, $startChars);
+                                $endPos = new Position($lineNo - 1, $endChars);
+                            }
+
+                            $diagnostics[] = Diagnostic::fromArray([
+                                'message' => yield $this->explainRule($rule),
+                                'range' => new Range($startPos, $endPos),
+                                'severity' => DiagnosticSeverity::WARNING,
+                                'source' => $this->name().':'.$rule
```

Comment:
should we be "contaminating" the source with the rule name? what if it's used to surpress diagnostics on the client side?

Reply:
![image](https://user-images.githubusercontent.com/2505114/215478925-ce32a635-fad7-4480-ab9f-f710db0b4783.png)

This is how that could look with client I use. For this PR this looks exactly same, but without the link